ksmt-bitwuzla/src/main/kotlin/org/ksmt/solver/bitwuzla/KBitwuzlaContext.kt

```diff
@@ -248,4 +295,82 @@ open class KBitwuzlaContext : AutoCloseable {
         override fun visit(sort: KFpRoundingModeSort) {
         }
     }
+
+    /**
+     * Move expressions from other cache level to the current cache level
+     * and register declarations for all moved expressions.
+     *
+     * 1. If expression is valid on previous levels we don't need to move it.
+     * Also, all expression declarations are known.
+     *
+     * 2. Otherwise, move expression to current level and recollect declarations.
+     * */
+    private inner class ExprMover : KNonRecursiveTransformer(ctx) {
+        override fun <T : KSort> transformExpr(expr: KExpr<T>): KExpr<T> {
+            // Move expr to current level
+            val term = exprGlobalCache.getValue(expr)
+            exprCacheLevel[expr] = currentLevel
+            exprCurrentLevelCache[expr] = term
+
+            return super.transformExpr(expr)
+        }
+
+        override fun <T : KSort> exprTransformationRequired(expr: KExpr<T>): Boolean {
+            val cachedLevel = exprCacheLevel[expr]
+            if (cachedLevel != null && cachedLevel < currentLevel) {
+                val levelCache = exprLeveledCache[cachedLevel]
+                // If expr is valid on its level we don't need to move it
+                return expr !in levelCache
+            }
+            return super.exprTransformationRequired(expr)
+        }
+
+        private var currentlyIgnoredDeclarations: Set<KDecl<*>>? = null
+
+        private fun registerDeclIfNotIgnored(decl: KDecl<*>) {
+            if (currentlyIgnoredDeclarations?.contains(decl) == true) {
+                return
+            }
+            registerDeclaration(decl)
+        }
+
+        override fun <T : KSort> transform(expr: KFunctionApp<T>): KExpr<T> {
+            registerDeclIfNotIgnored(expr.decl)
+            return super.transform(expr)
+        }
+
+        override fun <T : KSort> transform(expr: KConst<T>): KExpr<T> {
+            registerDeclIfNotIgnored(expr.decl)
+            return super.transform(expr)
+        }
+
+        override fun <D : KSort, R : KSort> transform(expr: KFunctionAsArray<D, R>): KExpr<KArraySort<D, R>> {
+            registerDeclIfNotIgnored(expr.function)
+            return super.transform(expr)
+        }
+
+        private val quantifiedVarsScope = arrayListOf<Pair<KExpr<*>, Set<KDecl<*>>?>>()
+
+        private fun <T : KSort> KExpr<T>.transformQuantifier(bounds: List<KDecl<*>>, body: KExpr<*>): KExpr<T> {
+            if (quantifiedVarsScope.lastOrNull()?.first != this) {
+                quantifiedVarsScope.add(this to currentlyIgnoredDeclarations)
+                val ignoredDecls = currentlyIgnoredDeclarations?.toMutableSet() ?: hashSetOf()
```

Comment:
I think, if you create `hashSetOf` to avoid redundant `Linked` versions, it is better to use `toHashSet` here